### DISCUSSION TITLE
How to test onDispose in tests?

### DISCUSSION THREAD

**Discussion by vixez** [rrousselGit/riverpod#3505] [Q&A] [answered]
Hello,

I'm using Dio for API calls, and would like to test if the `CancelToken` works when a provider is being disposed.
For example, I can add an `onDispose` and pass the `CancelToken` along to Dio:
```
CancelToken disposeCancelToken(Ref ref) {
  final cancelToken = CancelToken();
  ref.onDispose(() => cancelToken.cancel());
  return cancelToken;
}

CancelToken cancelToken = disposeCancelToken(-- ref would come here ---);

await expectLater(
      container.read(getTranslationsForAppProvider(
        appId: translationsAppId,
        appVersion: TestSetup.kAppVersion,
        country: 'BE',
        locale: 'nl',
        cancelToken: CancelToken(),
      ).future),
      completion(predicate<TranslationsResponseModel>((v) {
        return cancelToken.isCancelled;
  }, 'Check CancelToken got cancelled')),
);
```

And then I can check `cancelToken.isCancelled`.

But my questions would be:

1. How would I listen to `onDispose` from `ProviderContainer` in my integration test? (This is also kind of, how do you get the `Ref` to listen to `onDispose`?)
2. How can I trigger the onDispose immediately after reading the provider, so `onDispose` would get called?

Thanks in advance!

**Comment by rrousselGit** [comment, marked as answer]
> 3.. How can I trigger the onDispose immediately after reading the provider, so onDispose would get called?

Call `container.invalidate(provider)`. That should do the trick

> How would I listen to onDispose from ProviderContainer in my integration test? (This is also kind of, how do you get the Ref to listen to onDispose?)

Depends on the scenario. For testing your `disposeCancelToken` utility, you can make a new provider:

```dart
test('calls .cancel', () {
  late CancelToken token;
  final provider = Provider((ref) {
    token = disposeCancelToken(ref);
  });
  final container = ...


  container.read(provider);
  expect(token.isCancelled, false);

  container.invalidate(provider);

  expect(token.isCancelled, true);
})
```

For testing that a specific provider internally uses your `disposeCancelToken`, you'll have to find an indirect way to test this.

For instance, rather than testing "CancelToken.cancel" was called, you can test "After the provider is disposed, the previous `.future` throws a CancelledException"

For example:

```dart
test('', () async {
  final container = ...;
  
  final future = container.read(provider.future);

  container.dispose();

  expect(future, throwsA(isA<CancelledException>());
}); 
```

**Reply by vixez** [reply]
Yes, this works.
Thank you very much!
